module4/gift-certificate-system/api/src/main/java/com/epam/esm/controller/OrderController.java

```diff
@@ -0,0 +1,160 @@
+import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.methodOn;
+
+/**
+ * @author Sergei Kristev
+ * <p>
+ * Gets data from rest in JSON format on path "/gift-certificates".
+ */
+@RestController
+@RequestMapping("/gift-certificates")
+public class OrderController {
+
+    private final OrderService orderService;
+
+    /**
+     * Accepts service layer objects.
+     *
+     * @param orderService OrderService instance.
+     */
+    @Autowired
+    public OrderController(OrderService orderService) {
+        this.orderService = orderService;
+    }
+
+    /**
+     * Gets list of all orders.
+     *
+     * @param page     page's number
+     * @param pageSize page size
+     * @return Orders list.
+     */
+    @ResponseStatus(HttpStatus.OK)
+    @GetMapping(value = "/orders", produces = {"application/hal+json"})
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
+    public CollectionModel<OrderDTO> findAllOrders(@RequestParam(value = "page") Optional<Integer> page,
+                                                @RequestParam(value = "page_size") Optional<Integer> pageSize) {
+        int pageNumber = page.orElse(DEFAULT_PAGE_NUMBER);
+        int pageSizeNumber = pageSize.orElse(DEFAULT_PAGE_SIZE);
+        ValidationUtils.checkPaginationData(pageNumber, pageSizeNumber);
+
+        Page<OrderDTO> orderList = orderService.getAllOrders(pageNumber-1, pageSizeNumber);
+
+        for (OrderDTO orderDTO : orderList) {
+            Link selfLink = linkTo(methodOn(OrderController.class)
+                    .findOrderById(orderDTO.getId())).withRel("currentOrder");
+            orderDTO.add(selfLink);
+        }
+
+        Link orders = linkTo(OrderController.class).slash("orders").withRel("ordersList");
+
+        CollectionModel<OrderDTO> collectionModel = new CollectionModel(orderList, orders);
+        if (pageNumber > 1) {
+            Link previousPage = linkTo(methodOn(OrderController.class)
+                    .findAllOrders(Optional.of(pageNumber - 1), Optional.of(pageSizeNumber))).withRel("previousPage");
+            collectionModel.add(previousPage);
+        }
+        if (pageNumber < orderList.getTotalPages()) {
+            Link nextPage = linkTo(methodOn(OrderController.class)
+                    .findAllOrders(Optional.of(pageNumber + 1), Optional.of(pageSizeNumber))).withRel("nextPage");
+            collectionModel.add(nextPage);
+        }
+        return collectionModel;
+    }
+
+    /**
+     * Gets order by id.
+     *
+     * @param id Order id.
+     * @return Order instance.
+     */
+    @ResponseStatus(HttpStatus.OK)
+    @GetMapping(value = "/orders/{id}", produces = {"application/hal+json"})
+    @PreAuthorize("hasAnyRole('ROLE_USER', 'ROLE_ADMIN') && @authorizationComponentImpl.isUsersOrder(principal, #id)")
+    public OrderDTO findOrderById(@PathVariable Long id) {
+        OrderDTO orderDTO = orderService.getOrderById(id);
+        return addHateoasLinksToOrder(orderDTO);
```

Comment:
This is a good idea to move HATEOAS related logic to separate method. Code looks clean and readable.